main/modes/mode_manual.c

```diff
@@ -0,0 +1,400 @@
+    if (manual->curPage == NULL)// || ((page_t*)manual->curPage->val)->chapter != manual->chapter)
+    {
+        manual->curPage = paginateText(manual->markdown, &manual->pages);
+
+        if (reverse)
+        {
+            manual->curPage = manual->pages.last;
+        }
+    }
+}
+
+void manualEnterMode(display_t* disp)
+{
+    manual = calloc(1, sizeof(manual_t));
+
+    manual->disp = disp;
+
+    manual->fgColor = c555;
+    manual->bgColor = c000;
+
+    manual->navigate = RELOAD_PAGE;
+    manual->reload = true;
+
+    loadFont(MANUAL_FONT_UI, &manual->uiFont);
+    manual->bodyFont = manual->uiFont;
+    loadFont(MANUAL_FONT_HEADER, &manual->headerFont);
+
+    loadWsg("arrow12.wsg", &manual->arrowWsg);
+
+    markdownParams_t params =
+    {
+        .xMin = MANUAL_SIDE_MARGIN, .yMin = MANUAL_TOP_MARGIN,
+        .xMax = manual->disp->w - MANUAL_SIDE_MARGIN, .yMax = manual->disp->h - MANUAL_BOTTOM_MARGIN,
+        .bodyFont = &manual->bodyFont,
+        .headerFont = &manual->headerFont,
+        .color = manual->fgColor,
+    };
+
+    memcpy(&manual->mdParams, &params, sizeof(markdownParams_t));
+
+    manual->chapter = chapters;
+}
+
+void manualExitMode(void)
+{
+    freeFont(&manual->bodyFont);
+    freeFont(&manual->headerFont);
+
+    freeWsg(&manual->arrowWsg);
+
+    page_t* page;
+    while (NULL != (page = pop(&manual->pages)))
+    {
+        free(page->mdPos);
+        free(page);
+    }
+    // maybe redundant but oh well
+    clear(&manual->pages);
+
+    if (manual->markdown != NULL)
+    {
+        freeMarkdown(manual->markdown);
+    }
+
+    if (manual->text != NULL)
+    {
+        free(manual->text);
+    }
+    free(manual);
+}
+
+void manualMainLoop(int64_t elapsedUs)
+{
+#define hasNext (manual->curPage->next != NULL || manual->chapter < lastChapter)
+#define hasPrev (manual->curPage->prev != NULL)
+
+    if (manual->navigate != NONE)
+    {
+        if (manual->reload)
+        {
+            if (manual->navigate == NEXT_PAGE)
+            {
+                manualNextPage();
+            }
+            else if (manual->navigate == PREV_PAGE)
+            {
+                manualPrevPage();
+            }
+            else if (manual->navigate == RELOAD_PAGE)
+            {
+                manualLoadText(false);
+            }
+
+            // Clear the screen
+            fillDisplayArea(manual->disp, 0, 0, manual->disp->w, manual->disp->h, manual->bgColor);
+
+            page_t* curPage = manual->curPage->val;
+            drawMarkdown(manual->disp, manual->markdown, &manual->mdParams, &curPage->mdPos, false);
+
+            manual->navigate = NONE;
+            manual->reload = false;
+        }
+        else
+        {
+            if (manual->navigate == RELOAD_PAGE
+                || (manual->navigate == NEXT_PAGE && hasNext)
+                || (manual->navigate == PREV_PAGE && hasPrev))
+            {
+                // Just draw the waiting icon and set us up to reload on the next frame
```

Comment:
How slow are operations that you need a waiting icon?

Reply:
So parsing is the slowest part and it can take a few seconds for longer "chapters." Pagination isn't quite as slow and it only ever happens once per chapter. Rendering takes about half a second or less, so it's about on par with an old kindle. There's a lot of room for pretty straightforward optimizations though, as I know there's at least a couple places where there are redundant calls to `drawTextWordWrap()` that can be removed without much effort.

Reply:
It also seems like we could cache parsed data if we have enough SPIRAM to do so

Reply:
@maxwelllc figured it out! Turns out it was all the logging going on that was slowing everything down. Put all the debug logging behind `#ifdef DEBUG` statements (left the error logging alone) and bam! No more need for caching, optimizations, or even a loading symbol.